### ISSUE TITLE
Multiprocessing and Pieces in Maxout Layer

### ISSUE THREAD

**Issue by colinglaes** [aigamedev/scikit-neuralnetwork#124]
Hey All,

The problem i am trying to solve is binary classification, I have 43 inputs, one being an integer, all others being an integer  (price range), the other being dummy variables.  My hidden layer has 30 neurons.  I have two questions below, please let me know if you have any questions or need me to elaborate at all!

1. Can I make the library use multiple cores?  I ran training for a dataset and saw that my peak CPU usage was only 30% and averaged around 15%, I would like to leverage more of my CPU.  
2. Can anyone elaborate on the Piecewise option for Layer specification?

My code is below

```python
from sknn.platform import gpu64
from sknn.platform import cpu64
from pandas import *
import pandas as pd
import numpy as np
import os, sys
from os.path import isfile, join
reload(sys)
sys.setdefaultencoding('utf8')
from sknn.mlp import Classifier, Layer
import multiprocessing

# Specify Input Data
input_path = str('/Users/colinglaes/Documents/')
input_file = str('test.csv')
input_complete = os.path.join(input_path, input_file)

# Read Data
data_set = pd.read_csv(input_complete, sep = '|', header = 0)

# Remove Index Column and Drop NA Records
data_set = data_set.drop('IX', axis = 1)
data_set = data_set.dropna()

# Replicate Records and Remove Records Column
data_set = data_set.loc[np.repeat(data_set.index.values,data_set['RECORDS'])]
data_set = data_set.drop('RECORDS', axis = 1)

# Create Neural Network
nn = Classifier(layers = [Layer('Maxout',  units = 30, pieces = 15),
			               Layer('Softmax', units = 2)],
				learning_rate = 0.00000001,
				valid_size = 0.25,
				f_stable = .001,
				n_iter = 20,
				verbose = True)

# Split X + Y
dataframe_y = data_set['TARGET']
data_set = data_set.drop('TARGET', axis = 1)
dataframe_x = data_set

# Convert to Numpy Arrays and Remove NA
array_y = dataframe_y.as_matrix()
array_x = dataframe_x.as_matrix()

# Create Training Data
train_covariate = array_x[0:len(array_x) / 2]
train_target = array_y[0:len(array_y) / 2]

# Create Test Data
test_covariate = array_x[len(array_x) / 2 : len(array_x)]
test_target = array_y[len(array_y) / 2 : len(array_y)]

# Fit Neural Network
nn.fit(train_covariate, train_target)

# Get Probability Results from Testing Data
test_prediction = nn.predict_proba(test_covariate)

# Grab Column Names
covariate_column_names = list(dataframe_x.columns.values)
target_column_names = ['TARGET']
prediction_column_names = ['PROBABILITY_1', 'PROBABILITY_2']

# Create Dataframes
covariates_df = pd.DataFrame(data = test_covariates, columns = covariates_column_names)
target_df = pd.DataFrame(data = test_target, columns = target_column_names)
prediction_df = pd.DataFrame(data = test_prediction, columns = prediction_column_names)

# Merge Data
output_df = pd.concat([covariates_df, target_df, prediction_df], axis = 1)

# Save Data
output_df.to_csv('/Users/colinglaes/Desktop/output.csv')
```

Thanks,
Colin

**Comment by pGit1** [comment]
Colin to answer question number 2, the piecewise parameter is to control
how many linear equations the hidden node syphons through when looking
for a maxium value and is ONLY applicalbe when your layer type is set to
"maxout" units.
It is easiest to understand this in the contex of a rectifier units. The
activation of a rectifier unit is equivalent to the maximum value between
the linear combination of the vector of input weights (w) and the vector of
features (x) plus a bias feeding into a specific hidden neuron. In other
words Rectifier Activation = max(0, wx+b). If wx+b is greater that is the
output of the neuron else the output of the neuron is 0. Make sense?

A maxout activation is basically like a rctifier but has more than ONE
"wx+b" to check through. So to write this in butchered form the activation
of a maxout unit with *THREE pieces* is equivalent to max(0, wx+b, w`x+b`,
w``x+b``). So a rectifer is just a special case of Maxout where pieces
equals one. Make sense? Note that the other linear combinations will have
different weights whose parameters are learned during training. This type
of activation is best coupled with Dropout training which was its original
purpose and intent when created.

Hope that helps!! :D

**Comment by alexjc** [comment]
We'll probably deprecate Maxout in future releases, switching away from PyLearn2. You can get similar results with Rectifier units (and variations there of).

For maxing out your CPU, you may need to use multiple processes and for example letting scikit-learn train multiple models at once (e.g. grid search).  I'm not sure if Theano supports multithreaded execution, but if it does you can use it from sknn too (must check the documentation).

**Comment by alexjc** [comment]
It's not trivial to use multiprocessing for this, just using `n_jobs=8` on scikit-learn random searches seemed to crash the child processes. I'm not sure why... Will open a separate ticket for this.
